### PR TITLE
Fft

### PR DESCRIPTION
**Pull request by Jammy2211**

### CODE REVIEW

File: autoarray/dataset/abstract_dataset.py [hunk 1]
```diff
@@ -40,7 +40,9 @@ def load(cls, filename) -> "AbstractDataset":
         with open(filename, "rb") as f:
             return pickle.load(f)
 
-    def __init__(self, data, noise_map, exposure_time_map=None, name=None):
+    def __init__(
+        self, data, noise_map, exposure_time_map=None, name=None, metadata=None
```
Comment (rhayes777) [inline autoarray/dataset/abstract_dataset.py]: Add type hints and put arguments over new lines

File: autoarray/operators/transformer.py [hunk 2]
```diff
@@ -1,28 +1,33 @@
...
-        self.grid_radians = grid_radians.in_1d_binned
+        self.grid = grid.in_1d_binned
 
         self.total_visibilities = uv_wavelengths.shape[0]
-        self.total_image_pixels = grid_radians.shape_1d
+        self.total_image_pixels = grid.shape_1d
 
         self.preload_transform = preload_transform
 
         if preload_transform:
```
Comment (rhayes777) [inline autoarray/operators/transformer.py]: This kind of thing might be better done using lazy instantiation

File: autoarray/operators/transformer.py [hunk 3]
```diff
@@ -112,3 +117,200 @@ def transformed_mapping_matrices_from_mapping_matrix(self, mapping_matrix):
         )
 
         return [real_transformed_mapping_matrix, imag_transformed_mapping_matrix]
+
+
+class TransformerFFT(object):
+    def __init__(self, uv_wavelengths, grid):
+
+        super(TransformerFFT, self).__init__()
```
Comment (rhayes777) [inline autoarray/operators/transformer.py]: You can just call super().__init__() in python3

File: autoarray/operators/transformer.py [hunk 4]
```diff
@@ -112,3 +117,200 @@ def transformed_mapping_matrices_from_mapping_matrix(self, mapping_matrix):
         )
 
         return [real_transformed_mapping_matrix, imag_transformed_mapping_matrix]
+
+
+class TransformerFFT(object):
```
Comment (rhayes777) [inline autoarray/operators/transformer.py]: No need to inherit from object in python 3

File: autoarray/operators/transformer.py [hunk 5]
```diff
@@ -112,3 +117,200 @@ def transformed_mapping_matrices_from_mapping_matrix(self, mapping_matrix):
...
+    def __init__(self, uv_wavelengths, grid):
+
+        super(TransformerFFT, self).__init__()
+
+        self.uv_wavelengths = uv_wavelengths.astype("float")
+        self.grid = grid.in_1d_binned
+
+        self.u_fft = np.fft.fftshift(
+            np.fft.fftfreq(
+                grid.shape_2d[0], grid.pixel_scales[0] * units.arcsec.to(units.rad)
```
Comment (rhayes777) [inline autoarray/operators/transformer.py]: self.u_fft, self.v_fft = [... for i in (1, 2)]

File: autoarray/operators/transformer.py [hunk 6]
```diff
@@ -112,3 +117,200 @@ def transformed_mapping_matrices_from_mapping_matrix(self, mapping_matrix):
...
+        u_fft_meshgrid, v_fft_meshgrid = np.meshgrid(self.u_fft, self.v_fft)
+
+        # ... This is nessesary due to the way the grid in autolens is set up.
+        self.shift = np.exp(
+            -2.0
+            * np.pi
+            * 1j
+            * (
+                self.grid.pixel_scales[0]
+                / 2.0
```
Comment (rhayes777) [inline autoarray/operators/transformer.py]: Maybe break this equation up?

File: autoarray/operators/transformer.py [hunk 7]
```diff
@@ -112,3 +117,200 @@ def transformed_mapping_matrices_from_mapping_matrix(self, mapping_matrix):
...
+                * units.arcsec.to(units.rad)
+                * u_fft_meshgrid
+                + self.grid.pixel_scales[0]
+                / 2.0
+                * units.arcsec.to(units.rad)
+                * v_fft_meshgrid
+            )
+        )
+
+        self.uv = np.array(
```
Comment (rhayes777) [inline autoarray/operators/transformer.py]: I wonder if array is happy to accept an iterable meaning you don't need to cast to list?

File: autoarray/operators/transformer.py [hunk 8]
```diff
@@ -112,3 +117,200 @@ def transformed_mapping_matrices_from_mapping_matrix(self, mapping_matrix):
...
+            list(zip(self.uv_wavelengths[:, 0], self.uv_wavelengths[:, 1]))
+        )
+
+    def visibilities_from_image(self, image):
```
Comment (rhayes777) [inline autoarray/operators/transformer.py]: Add a type hint

File: autoarray/operators/transformer.py [hunk 9]
```diff
@@ -112,3 +117,200 @@ def transformed_mapping_matrices_from_mapping_matrix(self, mapping_matrix):
...
+        if len(image.shape) != 2:
+            raise exc.ArrayException("Transformer image must be 2D")
+
+        # NOTE: The input image is flipped to account for the way autolens is generating images
+        z_fft = np.fft.fftshift(np.fft.fft2(np.fft.fftshift(image[::-1, :])))
+
+        # ...
+        z_fft_shifted = z_fft * self.shift
+
+        # ...
```
Comment (rhayes777) [inline autoarray/operators/transformer.py]: Errr?

File: autoarray/operators/transformer.py [hunk 10]
```diff
@@ -112,3 +117,200 @@ def transformed_mapping_matrices_from_mapping_matrix(self, mapping_matrix):
...
+        real_visibilities = real_interp(self.uv)
+        imag_visibilities = imag_interp(self.uv)
+
+        return vis.Visibilities(
+            visibilities_1d=np.stack((real_visibilities, imag_visibilities), axis=-1)
+        )
+
+    def transformed_mapping_matrices_from_mapping_matrix(self, mapping_matrix):
+        """
+        ...
```
Comment (rhayes777) [inline autoarray/operators/transformer.py]: Docs

File: autoarray/operators/transformer.py [hunk 11]
```diff
@@ -112,3 +117,200 @@ def transformed_mapping_matrices_from_mapping_matrix(self, mapping_matrix):
...
+            real_transfomed_mapping_matrix[:, source_pixel_1d_index] = visibilities.real
+            imag_transfomed_mapping_matrix[:, source_pixel_1d_index] = visibilities.imag
+
+        return [real_transfomed_mapping_matrix, imag_transfomed_mapping_matrix]
+
+
+class TransformerNUFFT(NUFFT_cpu):
+    def __init__(self, uv_wavelengths, grid):
+
+        super(TransformerNUFFT, self).__init__()
```
Comment (rhayes777) [inline autoarray/operators/transformer.py]: Docs, no need to pass super args

File: autoarray/operators/transformer.py [hunk 12]
```diff
@@ -112,3 +117,200 @@ def transformed_mapping_matrices_from_mapping_matrix(self, mapping_matrix):
...
+                + self.grid.pixel_scales[0]
+                / 2.0
+                * units.arcsec.to(units.rad)
+                * self.uv_wavelengths[:, 0]
+            )
+        )
+
+    def initialize_plan(self, ratio=2, interpolation_kernel=(6, 6)):
+
+        if not isinstance(ratio, int):
```
Comment (rhayes777) [inline autoarray/operators/transformer.py]: Why is this necessary?

**Comment by rhayes777** [comment]
You're supposed to fix the issues before you merge it!

**Comment by Jammy2211** [comment]
> You're supposed to fix the issues before you merge it!

Ah, the thing at the bottom said approved, never saw it does it on a commit by commit basis.

This transformer stuff is scientifically untested so I dont want to spend too much time on documenting / cleaning the code currently. We could be using a completely different library in a months time.
